### PR TITLE
Add inlet/outlet BCs for the 2D-SWE

### PR DESCRIPTION
**Pull request by patrickersing**
This PR implements inlet/outlet boundary conditions for the `ShallowWaterEquationsWetDry2D`.
The new boundary conditions `BoundaryConditionWaterHeight` and `BoundaryConditionMomentum` specify the water height or the momentum on the boundary and extrapolate the remaining quantity from the Riemann Invariant.

### CODE REVIEW

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 1]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+
+    # Evaluate the conservative flux at the boundary
+    flux = Trixi.flux(u_boundary, orientation, equations)
+
+    # Return the conservative and nonconservative fluxes. 
+    # The nonconservative part is zero as we assume a constant bottom topography at the boundary.
+    return (flux, zero(u_inner))
+end
+
+# Version for `StructuredMesh`, `UnstructuredMesh2D`, and `P4estMesh`
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:335]: This might not work on `StructuredMesh` because it doesn't treat all the normals as outward pointing (see [here](https://github.com/trixi-framework/Trixi.jl/blob/main/src/equations/compressible_euler_2d.jl#L366)). It should be fine for `UnstructuredMesh2D` and `P4estMesh`

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 2]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+
+    # Return the conservative and nonconservative fluxes. The nonconservative part is zero as we assume
+    # a constant bottom topography at the boundary.
+    return (flux, zero(u_inner))
+end
+
+"""
+    BoundaryConditionMomentum(hv1_boundary, hv2_boundary, equations::ShallowWaterEquationsWetDry2D)
+
+Create a boundary condition that sets a fixed momentum in x- and y- direction, `hv1_boundary`
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:382]:
```suggestion
Create a boundary condition that sets a fixed momentum in x- and y- directions, `hv1_boundary`
```

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 3]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+    return (flux, zero(u_inner))
+end
+
+# Version for `StructuredMesh`, `UnstructuredMesh2D`, and `P4estMesh`
+function (boundary_condition::BoundaryConditionWaterHeight)(u_inner,
+                                                            normal_direction,
+                                                            x, t,
+                                                            surface_flux_functions,
+                                                            equations::ShallowWaterEquationsWetDry2D)
+    # Extract the  gravitational acceleration
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:341]:
```suggestion
    # Extract the gravitational acceleration
```

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 4]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+    return BoundaryConditionMomentum(t -> (hv1_boundary(t), hv2_boundary(t)))
+end
+
+# Version for `TreeMesh`
+function (boundary_condition::BoundaryConditionMomentum)(u_inner,
+                                                         orientation,
+                                                         direction, x, t,
+                                                         surface_flux_functions,
+                                                         equations::ShallowWaterEquationsWetDry2D)
+    # Extract the  gravitational acceleration
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:425]:
```suggestion
    # Extract the gravitational acceleration
```

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 5]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+    h_inner = waterheight(u_inner, equations)
+    v1_inner, v2_inner = velocity(u_inner, equations)
+
+    # Extract the external momentum from the boundary condition
+    hv1_boundary, hv2_boundary = boundary_condition.hv_boundary(t)
+
+    # Calculate the boundary state based on the direction.
+    # To extrapolate the external water height `h_boundary` assume that the Riemann invariant remains 
+    # constant across the incoming characteristic. This requires to solve for the roots of a nonlinear function.
+    # For convenience we substitute x = h_boundary and solve for x.
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:438]:
```suggestion
    # constant across the incoming characteristic. 
    # Requires one to solve for the roots of a nonlinear function, see Eq. (52) in the reference above.
    # For convenience we substitute x = h_boundary and solve for x.
```

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 6]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+
+    # Evaluate the conservative flux at the boundary
+    flux = Trixi.flux(u_boundary, orientation, equations)
+
+    # Return the conservative and nonconservative fluxes. 
+    # The nonconservative part is zero as we assume a constant bottom topography at the boundary.
+    return (flux, zero(u_inner))
+end
+
+# Version for `StructuredMesh`, `UnstructuredMesh2D` and `P4estMesh`
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:489]: Similar to comment above. This might not work on `StructuredMesh` due to orientation of the normals.

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 7]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+function (boundary_condition::BoundaryConditionMomentum)(u_inner,
+                                                         normal_direction,
+                                                         x, t,
+                                                         surface_flux_functions,
+                                                         equations::ShallowWaterEquationsWetDry2D)
+
+    # Extract the  gravitational acceleration
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:496]:
```suggestion
    # Extract the gravitational acceleration
```

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 8]
```diff
@@ -242,6 +241,302 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+    # Extract the external momentum from the boundary condition
+    hv1_boundary, hv2_boundary = boundary_condition.hv_boundary(t)
+
+    hv_boundary_normal = hv1_boundary * normal[1] + hv2_boundary * normal[2]
+    hv_boundary_tangential = -hv1_boundary * normal[2] + hv2_boundary * normal[1]
+
+    # Calculate the boundary state in the rotated coordinate system.
+    # To extrapolate the external water height `h_boundary` assume that the Riemann invariant remains 
+    # constant across the incoming characteristic. This requires to solve for the roots of a nonlinear function.
+    # For convenience we substitute x = h_boundary and solve for x.
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:518]:
```suggestion
    # constant across the incoming characteristic. 
    # Requires one to solve for the roots of a nonlinear function, see Eq. (52) in the reference above.
    # For convenience we substitute x = h_boundary and solve for x.
```

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 9]
```diff
@@ -857,7 +1192,7 @@ end
 end
 
 @inline function Trixi.waterheight_pressure(u, equations::ShallowWaterEquationsWetDry2D)
-    return Trixi.waterheight(u, equations) * Trixi.pressure(u, equations)
+    return waterheight(u, equations) * Trixi.pressure(u, equations)
 end
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_2d.jl:1196]: The function `waterheight_pressure` is exported by Trixi (see [here](https://github.com/trixi-framework/Trixi.jl/blob/main/src/Trixi.jl#L235)). So this entire function may be unnecessary.

**Review by andrewwinters5000 — CHANGES_REQUESTED** [review]
Thanks for adding this flexible boundary condition capability! It should work directly in 1D, right? It would be good to add an example elixir on `TreeMesh` if that is the case.

**Comment by patrickersing** [comment]
> Thanks for adding this flexible boundary condition capability! It should work directly in 1D, right? It would be good to add an example elixir on `TreeMesh` if that is the case.

@andrewwinters5000  Yes, this should directly carry over to the 1D case. I only need to implement the boundary condition functions in `ShallowWaterEquationsWetDry1D`. I was planning to split this off into a separate PR, but I can of course also add this directly here.

File: examples/tree_2d_dgsem/elixir_shallowwater_inflow_outflow.jl [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+# semidiscretization of the shallow water equations to test inflow/outflow boundary conditions
+
+equations = ShallowWaterEquationsWetDry2D(gravity = 9.81)
+
+# Setup initial conditions for a smooth channel flow with constant water height and velocity
+function initial_condition_channel_flow(x, t, equations::ShallowWaterEquationsWetDry2D)
+    H = 1.0
+    v1 = -0.1
+    v2 = -0.1
+    b = 0.0
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_2d_dgsem/elixir_shallowwater_inflow_outflow.jl:15]: The boundary condition only assumes the bottom topography to be constant at the boundary. Inside the domain there should be no restriction on the bottom topography.

File: examples/unstructured_2d_dgsem/elixir_shallowwater_inflow_outflow.jl [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+###############################################################################
+# semidiscretization of the shallow water equations to test inflow/outflow boundary conditions
+
+equations = ShallowWaterEquationsWetDry2D(gravity = 9.81)
+
+# Setup initial conditions for a smooth channel flow with constant water height and velocity
+function initial_condition_channel_flow(x, t, equations::ShallowWaterEquationsWetDry2D)
+    H = 1.0
+    v1 = 0.4
+    v2 = 0.4
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/unstructured_2d_dgsem/elixir_shallowwater_inflow_outflow.jl:15]: As long as we pick the velocities to be the same the velocity vector will be parallel to the walls. The specific value `v = 1.0 / sqrt(2.0)` should just normalize the velocity vector, so I think the current setting should be fine.

File: test/test_unstructured_2d.jl [hunk 12]
```diff
@@ -347,6 +347,32 @@ isdir(outdir) && rm(outdir, recursive = true)
...
+                                0.17890511959110378,
+                                0.0
+                            ],
+                            linf=[
+                                0.10340479800696867,
+                                0.3998099631746912,
+                                0.4041504686758526,
+                                0.0
+                            ],
+                            tspan=(0.0, 3.0))
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply test/test_unstructured_2d.jl:366]: In this case I picked a longer time to capture the moment when the prescribed wave at the outlet reaches the inlet in the testing to ensure that the BC can handle this scenario.

File: src/equations/shallow_water_wet_dry_2d.jl (hunk 1 above)
Comment (patrickersing) [reply src/equations/shallow_water_wet_dry_2d.jl:335]: Oh, I haven't been aware of this. Then I will remove the `StructuredMesh` in the comment.

File: src/equations/shallow_water_wet_dry_2d.jl (hunk 9 above)
Comment (patrickersing) [reply src/equations/shallow_water_wet_dry_2d.jl:1196]: Yes, we could replace this. Since we might move the SWE from Trixi.jl in the near future anyways, I would just leave it like this if you don't mind.

**Comment by patrickersing** [comment]
There are some uncovered lines from the `inverse_transform` functions included in the elixirs. I would either keep it as it is and accept the uncovered lines, since they only occur in the elixir, or move the `inverse_transform` function to the `shallow_water_equations_wet_dry.1d` and add proper tests. However, since the function will only be used for these steady-state setups I am not sure if it makes sense to add the function to the code.

File: src/equations/shallow_water_wet_dry_1d.jl [hunk 13]
```diff
@@ -217,6 +217,159 @@ For details see Section 9.2.5 of the book:
...
+    # Calculate the boundary state based on the direction.
+    # To extrapolate the external water height `h_boundary` assume that the Riemann invariant remains 
+    # constant across the incoming characteristic. 
+    # Requires one to solve for the roots of a nonlinear function, see Eq. (52) in the reference above.
+    # For convenience we substitute x = h_boundary and solve for x.
+    if direction == 1 # x-
+        fx = ZeroProblem(x -> 2 * sqrt(g) * x^(3 / 2) +
+                              (v_inner - 2 * sqrt(g * h_inner)) * x - hv_boundary,
+                         h_inner)
+        h_boundary = solve(fx, Order2())
```
Comment (andrewwinters5000) [inline src/equations/shallow_water_wet_dry_1d.jl:355]: What does `Order2()` mean and / or do here? Perhaps a comment to indicate what this means

**Review by andrewwinters5000 — CHANGES_REQUESTED** [review]
Looks very nice, I just left some comments mainly regarding coding style.

File: examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+        D(f) = h_hat -> Trixi.ForwardDiff.derivative(f, float(h_hat))
+
+        # Solve the root finding problem using Newton's method
+        h_hat = Roots.newton((f, D(f)), h_hat_init)
+    else
+        throw(error("The given state is not admissible: E_hat = $E_hat, sigma = $sigma"))
+    end
+
+    # Return the water height `h = h_hat * h_0`.
+    return h_hat * h_0
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl:58]: `h_hat = h / h_0` is a normalized water height, so we have to rescale to obtain the actual water height. 
I have expanded the comment to explain this. Also looking through the reference, I now reference equation (2.19) instead of (2.9) in line 37, which should also make this more clear.

[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl:58]: h_hat = h / h_0 is a normalized water height, so we have to rescale to obtain the actual water height. I have expanded the comment to explain this. Also looking through the reference, I now reference equation (2.19) instead of (2.9) in line 37, which should also make this more clear.

File: examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+
+    # Set the total energy before and after the shock
+    if x[1] < 11.665504281554291
+        E = 3 / 2 * (g * 0.18)^(2 / 3) + g * 0.2
+    else
+        E = 0.18^2 / (2 * 0.33^2) + g * 0.33
+    end
+
+    # # Set the quadratic bottom topography function
+    if 8 ≤ x[1] ≤ 12
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl:87]: Sure, we can use the explicit version here. 

If we don't match the corner points exactly, the projection of the bottom topography will not be exact. 
Other than that I don't think any special treatment is necessary as the bottom topography is continuous in these test cases, so the left/right element values of the bottom topography are identical.

File: examples/tree_1d_dgsem/elixir_shallowwater_moving_water_subsonic.jl [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+    h_0 = 1 / g * (g * abs(hv))^(2 / 3)
+    phi_0 = 3 / 2 * (g * abs(hv))^(2 / 3)
+
+    # normalized total energy
+    E_hat = (E - g * b) / phi_0
+
+    # Check if the state is admissible and compute the water height from equation (2.9).
+    if (sigma == 0) && (E_hat ≈ 1) # sonic state
+        h_hat = 1
+    elseif abs(sigma) == 1 && E_hat > 1 # supersonic / subsonic state
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_subsonic.jl:37]: There is no specific reason for it. I removed the parenthesis as we also don't have them in other parts of the code.

File: examples/tree_1d_dgsem/elixir_shallowwater_moving_water_subsonic.jl [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+        D(f) = h_hat -> Trixi.ForwardDiff.derivative(f, float(h_hat))
+
+        # Solve the root finding problem using Newton's method
+        h_hat = Roots.newton((f, D(f)), h_hat_init)
+    else
+        throw(error("The given state is not admissible: E_hat = $E_hat, sigma = $sigma"))
+    end
+
+    # Return the water height `h = h_hat * h_0`.
+    return h_hat * h_0
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_subsonic.jl:56]: I have expanded the comment to explain this better.

File: examples/tree_1d_dgsem/elixir_shallowwater_moving_water_transonic.jl [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+        D(f) = h_hat -> Trixi.ForwardDiff.derivative(f, float(h_hat))
+
+        # Solve the root finding problem using Newton's method
+        h_hat = Roots.newton((f, D(f)), h_hat_init)
+    else
+        throw(error("The given state is not admissible: E_hat = $E_hat, sigma = $sigma"))
+    end
+
+    # Return the water height `h = h_hat * h_0`.
+    return h_hat * h_0
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_transonic.jl:57]: I have expanded the comment to explain this better.

File: src/equations/shallow_water_wet_dry_1d.jl (hunk 13 above)
Comment (patrickersing) [reply src/equations/shallow_water_wet_dry_1d.jl:355]: The function argument specifies the algorithm to solve the non-linear problem. In this case `Order2()` implements a version of the Steffensen method.

File: src/equations/shallow_water_wet_dry_2d.jl [hunk 19]
```diff
@@ -243,6 +242,304 @@ Should be used together with [`Trixi.TreeMesh`](@extref).
...
+    # Calculate the boundary state based on the direction.
+    # To extrapolate the external water height `h_boundary` assume that the Riemann invariant remains 
+    # constant across the incoming characteristic. 
+    # Requires one to solve for the roots of a nonlinear function, see Eq. (52) in the reference above.
+    # For convenience we substitute x = h_boundary and solve for x.
+    if direction == 1 # x-
+        fx = ZeroProblem(x -> 2 * sqrt(g) * x^(3 / 2) +
+                              (v1_inner - 2 * sqrt(g * h_inner)) * x - hv1_boundary,
+                         h_inner)
+        h_boundary = solve(fx, Order2())
```
[the remark this answers is not in the document]
Comment (patrickersing) [reply src/equations/shallow_water_wet_dry_2d.jl:445]: The function argument specifies the algorithm to solve the non-linear problem. In this case Order2() implements a version of the Steffensen method. I have added a respective comment to the top of the if statement.

File: examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl (hunk 14 above)
[the remark this answers is not in the document]
Comment (andrewwinters5000) [reply examples/tree_1d_dgsem/elixir_shallowwater_moving_water_shock.jl]: Thanks, the new comments help clarify this scaling.

**Review by andrewwinters5000 — APPROVED** [review]
LGTM!

**Comment by andrewwinters5000** [comment]
> There are some uncovered lines from the `inverse_transform` functions included in the elixirs. I would either keep it as it is and accept the uncovered lines, since they only occur in the elixir, or move the `inverse_transform` function to the `shallow_water_equations_wet_dry.1d` and add proper tests. However, since the function will only be used for these steady-state setups I am not sure if it makes sense to add the function to the code.

I agree that it makes more sense to leave the function definition of `inverse_transform` in the elixirs where it is needed for these special test case setups. Also, losing the (very) small coverage percentage is warranted in this case.
